### PR TITLE
[mcp23017] Workaround for input pins not being initialized

### PR DESCRIPTION
**Pull request by aogorek**
Fixed INPUT PIN support after changes in 2.3.0.1

After upgrade to OH2 ver 2.3.0.0 INPUT pins in [mcp23017] binding stopped to work.
I investigated a little, and it turned out, that "channelLinked" method is not called at all, so input pins cannot be initialized.
First I suspected that it is due to wrong order of loading configuration files - items are loaded before things - but in ThingLinkManager there is method receiveTypedEvent which should be called after ThingStatusInfoChangedEvent (INITIALIZING -> ONLINE).

The problem is - it is not called for my binding and I really do not know why. I changed log level to TRACE and noticed that event itself is generated, but this handler is not called - so channelLinked is not called - so input pins are not initialized - so binding is not working.
I've been searching through git history and I've noticed that there were some quite big changes in thing initialization - now it is asynchronous - if I understood changes correctly, and I suspect that somewhere there is a reason. There might be something with threads synchronization, but I'm not sure.

So, I created some fix. I am aware that is quite dirty workaround of a problem, but it works well form me and is confirmed working by other user here:
https://community.openhab.org/t/problem-with-new-mcp23017-binding-in-openhab-2-3/49270/10

I also started a thread here:
https://community.openhab.org/t/invalid-load-order-of-configuration-file-not-calling-channellinked/49269/2

@kaikreuzer I know You are quite busy now, but I think You may know a lot more about recent changes in thing intialization, so please take a look at this in a free moment.

I suggest to temporarily add this change to repository, because now INPUT pins mostly doesn't work.
If main problem will be found I'll restore previous version.

### CODE REVIEW

**Comment by kaikreuzer** [comment]
I am actually not sure whether it was ever guaranteed to have this method called for each linked channel. The JavaDoc states "This method is only called, if the thing has been initialized", which I would read as: Only channels that are linked (=link is created) after handler initialization are reported here, all others, where links existed before, won't see such a call.

I think your PR is therefore the correct thing to do: At initialization, iterate over all linked channels once and then listen to any further updates.

File: addons/binding/org.openhab.binding.mcp23017/src/main/java/org/openhab/binding/mcp23017/handler/Mcp23017Handler.java [hunk 1]
```diff
@@ -57,13 +58,10 @@
 
     public Mcp23017Handler(Thing thing) {
         super(thing);
-    }
-
-    @Override
-    public void initialize() {
         checkConfiguration();
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.mcp23017/src/main/java/org/openhab/binding/mcp23017/handler/Mcp23017Handler.java]: Why did you move these methods to the constructor? Imho, this should remain in `initialize()` as before.

Comment (aogorek) [reply addons/binding/org.openhab.binding.mcp23017/src/main/java/org/openhab/binding/mcp23017/handler/Mcp23017Handler.java]: I thought so too, but sometimes channelLinked is called before initialize, and it ends with:

``
java.lang.NullPointerException: null
        at org.openhab.binding.mcp23017.handler.Mcp23017Handler.handleOutputCommand(Mcp23017Handler.java:112) ~[?:?]
        at org.openhab.binding.mcp23017.handler.Mcp23017Handler.handleCommand(Mcp23017Handler.java:79) ~[?:?]
        at org.eclipse.smarthome.core.thing.binding.BaseThingHandler.channelLinked(BaseThingHandler.java:278) ~[?:?]
        at org.openhab.binding.mcp23017.handler.Mcp23017Handler.channelLinked(Mcp23017Handler.java:222) ~[?:?]
        at org.eclipse.smarthome.core.thing.link.ThingLinkManager.lambda$0(ThingLinkManager.java:300) ~[?:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:?]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) [?:?]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:?]
        at java.lang.Thread.run(Thread.java:745) [?:?]
``

its a little unpredictable for me now...

**Comment by kaikreuzer** [comment]
>  but sometimes channelLinked is called before initialize,

This cannot be the case. The exception you show goes through `ThingLinkManager.java:300`, which is only executed after a `ThingHandlerHelper.isHandlerInitialized(thing)` check - which means that the Thing status is either already set to ONLINE, OFFLINE or UNKNOWN and therefore the handler must have already been initialized.

**Comment by aogorek** [comment]
You are right. 
I verified a little more, added some more debugs to my binding  and it looks like it is called from 
Mcp23017Handler#initialize... It calls super.initialize, that calls  updateStatus(ThingStatus.ONLINE);
It probably fires ThingStatusInfoChangedEvent (INITIALIZING -> ONLINE) which I thought is not fired, and then channellLinked is executed and fails with NPE because 
pinStateHolder and mcpProvider are initialized in initialize() AFTER super.initialize().
Probably that's why it works when pinStateHolder and mcpProvider are initialized in constructor...

So what is Your suggestion? Should I keep these intializations in constructor? initialize method is marked as deprecated. Ofcourse  I could add some "if" checking in channelLinked and initialize in first method visit but IMHO it also would not be very clean code.

**Comment by kaikreuzer** [comment]
Ah, I guess [the answer is very simple](https://github.com/eclipse/smarthome/blob/00a7970fcb3a3b6d99dfee4055c7d245c9f57cb2/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandler.java#L196-L202): Do not call the super implementation!

**Comment by aogorek** [comment]
It looks like it was a case... :)
I've commited fixes
